File: 901.online-stock-span.go
/*
 * @lc app=leetcode id=901 lang=golang
 *
 * [901] Online Stock Span
 */

// @lc code=start
type Node struct {
    Val int
    res int
}

type StockSpanner struct {
    Spans []Node
}


func Constructor() StockSpanner {
    return StockSpanner{make([]Node, 0)}
}


func (this *StockSpanner) Next(price int) int {
    res := 1

    if len(this.Spans) == 0{
        this.Spans = append(this.Spans, Node{price, res})
        return res
    }

    for len(this.Spans) > 0 && this.Spans[len(this.Spans)-1].Val <= price{
        res += this.Spans[len(this.Spans) - 1].res
        this.Spans = this.Spans[:len(this.Spans) - 1]
    }

    this.Spans = append(this.Spans, Node{price, res})

    return res
}


/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */
// @lc code=end

File: 901.online-stock-span.cpp
/*
 * @lc app=leetcode id=901 lang=cpp
 *
 * [901] Online Stock Span
 */

// @lc code=start
class StockSpanner {
 public:
  StockSpanner() {}

  int next(int price) {
    int span = 1;
    while (!st_.empty() && price >= st_.top().first) {
      span += st_.top().second;
      st_.pop();
    }

    if (st_.empty() || price < st_.top().first) {
      st_.push({price, span});
    }

    return span;
  }

 private:
  stack<pair<int, int>> st_;  // price, span
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */
// @lc code=end

